File: ascii_fields/animations/magnetic.py
import math

from ..core import Animation, clamp, render_field


class MagneticAnimation(Animation):
  """The dipole field of a bar magnet -- just the field lines, no drawn magnet.
  Field lines arc from N to S and a current of tracers flows along them. The
  pattern is the streamfunction psi = angle(N) - angle(S); its level sets are
  the lines, and a gentle |B|-like falloff keeps them readable everywhere."""

  default_theme = "ice"

  thresholds = [
    (0.18, " "), (0.30, "."), (0.42, ":"), (0.54, "-"), (0.66, "="),
    (0.76, "+"), (0.85, "*"), (0.93, "#"), (1.01, "@"),
  ]

  def render(self, width, height, elapsed, phase, options):
    flow = elapsed * 2.2
    lines = 9.0
    d = 0.34
    ax = width / max(1, height * 2.0)
    contrast = options.contrast
    grid = []
    for row in range(height):
      py = (row / max(1, height - 1) - 0.5) * 2.0
      line = []
      for col in range(width):
        px = (col / max(1, width - 1) - 0.5) * 2.0 * ax
        # streamfunction of two opposite poles on the x-axis
        psi = math.atan2(py, px - d) - math.atan2(py, px + d)
        field_lines = 0.5 + 0.5 * math.sin(lines * psi - flow)
        # field strength: gentle |B|-like falloff, capped so the poles don't
        # paint a saturated 'bar magnet' blob between them
        rn = math.hypot(px - d, py) + 0.18
        rs = math.hypot(px + d, py) + 0.18
        strength = clamp(0.18 + 0.22 / rn + 0.22 / rs, 0.0, 0.95)
        # darken the immediate neighbourhood of each pole so they don't blob
        pole_cut = max(math.exp(-(rn * rn) * 18.0), math.exp(-(rs * rs) * 18.0))
        level = field_lines * strength * (1.0 - 0.85 * pole_cut)
        line.append(clamp(level * contrast))
      grid.append(line)
    return render_field(width, height, grid, options, self)

File: src/animations/magnetic.rs
use crate::animation::{Animation, FrameContext};
use crate::core::{clamp, render_field, FieldStyle};

const STYLE: FieldStyle = FieldStyle { gray_lo: 234, gray_hi: 255, default_theme: "ice" };
const TH: &[(f64, char)] = &[
  (0.18, ' '), (0.30, '.'), (0.42, ':'), (0.54, '-'), (0.66, '='),
  (0.76, '+'), (0.85, '*'), (0.93, '#'), (1.01, '@'),
];

pub struct Magnetic;

impl Animation for Magnetic {
  fn render(&mut self, ctx: &FrameContext, out: &mut String) {
    let w = ctx.width; let h = ctx.height;
    let mut grid = vec![0.0_f64; w * h];
    let t = ctx.elapsed;
    let flow = t * 2.2;
    let lines = 9.0_f64;
    let d = 0.34_f64;
    let ax = w as f64 / ((h as f64) * 2.0).max(1.0);
    let contrast = ctx.options.contrast;
    let dw = (w.saturating_sub(1)).max(1) as f64;
    let dh = (h.saturating_sub(1)).max(1) as f64;
    for row in 0..h {
      let py = (row as f64 / dh - 0.5) * 2.0;
      let py2 = py * py;
      let base = row * w;
      for col in 0..w {
        let px = (col as f64 / dw - 0.5) * 2.0 * ax;
        let psi = py.atan2(px - d) - py.atan2(px + d);
        let field_lines = 0.5 + 0.5 * (lines * psi - flow).sin();
        let rn = ((px - d).powi(2) + py2).sqrt() + 0.18;
        let rs = ((px + d).powi(2) + py2).sqrt() + 0.18;
        let strength = (0.18 + 0.22 / rn + 0.22 / rs).clamp(0.0, 0.95);
        let pole_cut = (-(rn * rn) * 18.0).exp().max((-(rs * rs) * 18.0).exp());
        let level = field_lines * strength * (1.0 - 0.85 * pole_cut);
        grid[base + col] = clamp(level * contrast);
      }
    }
    render_field(ctx, &grid, TH, &STYLE, out);
  }
}
